Status: Add microshift as part of status

praveenkumar opened the pull request:
With this patch now for microshift preset user will get following
output.
```
$ ./crc status
CRC VM:          Running
MicroShift:      Running (v4.12.9)
RAM Usage:       1.144GB of 3.913GB
Disk Usage:      4.647GB of 16.1GB (Inside the CRC VM)
Cache Usage:     208.9GB
Cache Directory: /home/prkumar/.crc/cache

$ ./crc status -ojson
{
  "success": true,
  "crcStatus": "Running",
  "microshiftStatus": "Running",
  "microshiftVersion": "4.12.9",
  "diskUsage": 4646580224,
  "diskSize": 16095641600,
  "cacheUsage": 208867411129,
  "cacheDir": "/home/prkumar/.crc/cache",
  "ramSize": 3912945664,
  "ramUsage": 1144094720,
  "preset": "microshift"
}
```

Fixes: #3592

praveenkumar committed:
Bundle: Check for base/app domain except for podman bundle

praveenkumar committed:
Start: Fix message if user start an already running instance

With this PR now if user start and running microshift instance then
it shows correct message.
```
 $ ./crc status
CRC VM:          Running
RAM Usage:       1.473GB of 3.913GB
Disk Usage:      6GB of 16.1GB (Inside the CRC VM)
Cache Usage:     208.9GB
Cache Directory: /home/prkumar/.crc/cache

$ ./crc start
INFO Loading bundle: crc_microshift_libvirt_4.12.9_amd64...
INFO A CRC VM for MicroShift 4.12.9 is already running
Started the MicroShift cluster.

export KUBECONFIG="/home/prkumar/.crc/machines/crc/kubeconfig"

Use the 'oc' command line interface:
  $ eval $(crc oc-env)
  $ oc COMMAND
```

Without this PR it was showing podman preset message template.
```
$ ./crc start
INFO Loading bundle: crc_microshift_libvirt_4.12.9_amd64...
INFO A CRC VM for Podman  is already running
podman runtime is now running.

Use the 'podman' command line interface:
  $ eval $(crc podman-env)
  $ podman-remote COMMAND
```

praveenkumar committed:
ClusterOperator: Rename kubernetesClient to openShiftClient

`kubernetesClient` function actually provide the openshiftClient.
Because this client doesn't have any core k8s methods like `AppsV1`,
`CoreV1` ..etc. and took me some time to figure out what's going
on.

evidolob reviewed (approved):
This work as expected:

```json
{"CrcStatus":"Running",
"OpenshiftStatus":"",
"MicroShiftStatus":"Running",
"MicroShiftVersion":"4.12.9",
"DiskUse":4739710976,
"DiskSize":16095641600,
"RAMUse":979648512,
"RAMSize":3913109504,
"Preset":"microshift"
}
```

anjannath commented on pkg/crc/machine/status.go:
```diff
@@ -57,13 +64,20 @@ func (client *client) Status() (*types.ClusterStatusResult, error) {
+		clusterStatusResult.PodmanVersion = vm.bundle.GetPodmanVersion()
+		clusterStatusResult.Preset = preset.Podman
+	case vm.bundle.IsMicroshift():
+		logging.Info("microshift")
```
i think this `log.Info` is unintended and left over from debugging

praveenkumar replied:
@anjannath ah yes, I will remove those.

praveenkumar replied:
@anjannath Updated.

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -295,13 +295,13 @@ func (client *client) Start(ctx context.Context, startConfig types.StartConfig)
-		if !vm.bundle.IsOpenShift() {
-			logging.Infof("A CRC VM for Podman %s is already running", vm.bundle.GetPodmanVersion())
+		if vm.bundle.IsPodman() {
+			logging.Infof("A CRC VM for %s %s is already running", startConfig.Preset.ForDisplay(), vm.bundle.GetPodmanVersion())
```
This can be changed to `logging.Infof("A CRC VM for %s %s is already running", startConfig.Preset.ForDisplay(), vm.bundle.GetVersion())` and be moved out of the `if {}` block.

```
  logging.Infof("A CRC VM for %s %s is already running", startConfig.Preset.ForDisplay(), vm.bundle.GetVersion())
  if vm.bundle.IsPodman() {
    return &types.StartResult{
				Status: vmState,
			}, nil
  }
  clusterConfig, err := getClusterConfig(vm.bundle)
  if err != nil {
    return nil, errors.Wrap(err, "Cannot create cluster configuration")
```

cfergeau commented on pkg/crc/cluster/clusteroperator.go:
```diff
@@ -80,7 +80,7 @@ func (status *Status) IsReady() bool {
 
 func GetClusterOperatorsStatus(ctx context.Context, ip string, kubeconfigFilePath string) (*Status, error) {
-	lister, err := kubernetesClient(ip, kubeconfigFilePath)
+	lister, err := openShiftClient(ip, kubeconfigFilePath)
```
fwiw, `openshiftClient` would be easier to read

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -132,6 +134,7 @@ func createBars(cpuUse []int64, writer io.Writer) (ramBar *pb.ProgressBar, cpuBa
 func getStatus(client *daemonclient.Client, cacheDir string) *status {
 
 	clusterStatus, err := client.APIClient.Status()
+
```
unneeded change

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -217,6 +225,13 @@ func openshiftStatus(status *status) string {
+	if status.MicroShiftVersion != "" {
+		return fmt.Sprintf("%s (v%s)", status.MicroShiftStatus, status.MicroShiftVersion)
+	}
+	return string(status.OpenShiftStatus)
```
MicroShiftStatus?

cfergeau commented on pkg/crc/machine/status.go:
```diff
@@ -146,6 +158,24 @@ func getOpenShiftStatus(ctx context.Context, ip string) types.OpenshiftStatus {
+	if microShiftServiceState == sdState.Running {
+		return types.OpenshiftRunning
+	}
+	return types.OpenshiftStopped
```
fwiw, `crc start` waits for `oc get nodes` to be successful before deciding the cluster was started successfully, maybe `status` should use the same criterion.

praveenkumar replied:
`crc start` wait for cluster operator status, for `oc get nodes` it just wait for API endpoint. so if microshift service is running that means API is already up.

cfergeau replied:
I am talking about `crc start` with the microshift preset, which only checks `oc get nodes` (this calls `waitForApiServer`)

praveenkumar replied:
Created a follow up https://github.com/crc-org/crc/issues/3599 on this.

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -39,19 +39,21 @@ var statusCmd = &cobra.Command{
+	OpenShiftVersion  string                       `json:"openshiftVersion,omitempty"`
+	MicroShiftStatus  types.OpenshiftStatus        `json:"microshiftStatus,omitempty"`
+	MicroShiftVersion string                       `json:"microshiftVersion,omitempty"`
+	PodmanVersion     string                       `json:"podmanVersion,omitempty"`
```
To be honest, this feels complex and error prone. `OpenShiftStatus`/`MicroShiftStatus` and `OpenShiftVersion`/`MicroShiftVersion` are mutually exclusive, and `OKDStatus`/`OKDVersion` would be missing. Maybe you could reuse `OpenShiftStatus`/`OpenShiftVersion` actually, and do something different when pretty-printing depending on the preset.
#3592 only mentions the `Preset` field being wrong, so we have some latitude with these fields.

praveenkumar replied:
it is not only around prettyprint but also json side.

cfergeau replied:
Even on the json side, if `Preset` is correctly set, I don't think we need redundant/mutually exclusive fields.

gbraad replied:
Additional fields (or those that change names), ahs to be checked in the receiving end for `undefined` or existence. Which also complicates the receiving end a lot.

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -217,6 +225,13 @@ func openshiftStatus(status *status) string {
 }
 
+func microShiftStatus(status *status) string {
+	if status.MicroShiftVersion != "" {
```
This is called as 
```
if s.MicroShiftVersion != "" {
		lines = append(lines, line{"MicroShift", microShiftStatus(s)})
}
```
so this test is unneeded. I would change the test above to 
```
if preset == Microshift {
		lines = append(lines, line{"MicroShift", microShiftStatus(s)})
}
```

praveenkumar replied:
yes I didn't change because same what we used for `Openshift` and `podman`, may be another commit to use the preset instead those.

cfergeau replied:
well, existing code is buggy, good opportunity to fix it.

gbraad commented on pkg/crc/api/client/types.go:
```diff
@@ -20,15 +20,17 @@ type StartResult struct {
+	OpenshiftVersion  string `json:"OpenshiftVersion,omitempty"`
+	MicroShiftStatus  string
+	MicroShiftVersion string `json:"MicroShiftVersion,omitempty"`
+	PodmanVersion     string `json:"PodmanVersion,omitempty"`
```
This becomes as mess on the receiving end.

Just a `Version` and `Status` will do, as the `Preset` is the key value to determine the meaning

gbraad reviewed (changes requested):
This will complicate the logic on both ends and maybe even impede future changes. Each new type of bundle/preset will need new fields?

```json
{
  "success": true,    // <!--- ???
  "crcStatus": "Running",
  "microshiftStatus": "Running",
  "microshiftVersion": "4.12.9",
  ...
  "preset": "microshift"
}
```
could be:

```json
{
  "crcStatus": "Running",
  ...
  "preset": "microshift"
  "status": "Running",
  "version": "4.12.9",
}
```

praveenkumar committed:
Status: Add microshift as part of status

With this patch now for microshift preset user will get following
output.
```
$ ./crc status
CRC VM:          Running
MicroShift:      Running (v4.12.9)
RAM Usage:       1.144GB of 3.913GB
Disk Usage:      4.647GB of 16.1GB (Inside the CRC VM)
Cache Usage:     208.9GB
Cache Directory: /home/prkumar/.crc/cache

$ ./crc status -ojson
{
  "success": true,
  "crcStatus": "Running",
  "openshiftStatus": "Running",
  "openshiftVersion": "4.12.9",
  "diskUsage": 4642684928,
  "diskSize": 16095641600,
  "cacheUsage": 208867411129,
  "cacheDir": "/home/prkumar/.crc/cache",
  "ramSize": 3912953856,
  "ramUsage": 1007919104,
  "preset": "microshift"
}

```

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -180,12 +180,15 @@ func (s *status) prettyPrintTo(writer io.Writer) error {
 		lines = append(lines, line{"OpenShift", openshiftStatus(s)})
 	}
-	if s.PodmanVersion != "" {
+	if s.Preset == preset.Podman {
```
With openshift/microshift, we also expose podman, so I think we always want to print this if podman version is set.

praveenkumar replied:
@cfergeau I discussed it with @gbraad about the same thing and he don't want to have podman version in case of openshift/microshift preset otherwise some user might think that this cluster running on podman.

cfergeau replied:
Nobody can know this except you and Gerard. Please split this behaviour change to its own commit with a proper rationale so that it's obvious to everyone that the change is intentional, and why it's being done.

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -217,6 +220,13 @@ func openshiftStatus(status *status) string {
 }
 
+func microShiftStatus(status *status) string {
+	if status.Preset == preset.Microshift {
```
Shouldn't this be `if status.OpenShiftVersion != ""`? This makes the method identical to `openshiftStatus`?

cfergeau commented on pkg/crc/machine/status.go:
```diff
@@ -57,13 +64,18 @@ func (client *client) Status() (*types.ClusterStatusResult, error) {
+		clusterStatusResult.OpenshiftStatus = getMicroShiftStatus(vm)
+		clusterStatusResult.OpenshiftVersion = vm.bundle.GetOpenshiftVersion()
+		clusterStatusResult.Preset = preset.Microshift
+	default:
```
The code is starting to get quite repetitive, can you first create a `clusterStatusResult` with with the fields Preset and xxxVersion set. Then you can use this variable in the 2 code paths (cluster is stopped or running) and only set the correct Status  field. This can even be moved to different helpers for improved readabililty.

praveenkumar committed:
Status: Use preset to show pretty print info

As of now if openshift preset is used then podman version
is also shown which is bit confusing for the user since they might
think that this cluster is running on podman. This PR make sure
only the preset related info should be part of status.

praveenkumar commented:
> This will complicate the logic on both ends and maybe even impede future changes. Each new type of bundle/preset will need new fields?

This should be now changed.

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -180,10 +180,10 @@ func (s *status) prettyPrintTo(writer io.Writer) error {
 		lines = append(lines, line{"OpenShift", openshiftStatus(s)})
 	}
-	if s.PodmanVersion != "" {
+	if s.Preset == preset.Podman {
```
This could become a `switch`.
